From 62840bc5191097366d6ed852e3af37ef46fe8d28 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 29 Dec 2016 12:04:12 -0800
Subject: [PATCH] CNS-190 AXFR failure due to very large number of records on a
 single name Reviewed by: Richard Kiene <richard.kiene@joyent.com>

---
 lib/dns-server.js | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/lib/dns-server.js b/lib/dns-server.js
index a69acba..bc9a23b 100644
--- a/lib/dns-server.js
+++ b/lib/dns-server.js
@@ -1160,15 +1160,20 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 					var name = names[i];
 					var recs = JSON.parse(zrecs[name]);
 					assert.arrayOfObject(recs);
-					self.addAnswers(q, recs,
-					    name + '.' + z);
-					pushed += recs.length;
-					if (pushed > MAX_ANS_PER_MSG) {
-						if (q.send() === false) {
-							finish();
-							return;
+					var trecs;
+					while (recs.length > 0) {
+						trecs = recs.slice(0, 20);
+						recs = recs.slice(20);
+						self.addAnswers(q, trecs,
+						    name + '.' + z);
+						pushed += trecs.length;
+						if (pushed > MAX_ANS_PER_MSG) {
+							if (!q.send()) {
+								finish();
+								return;
+							}
+							pushed = 0;
 						}
-						pushed = 0;
 					}
 				}
 				finish();
-- 
2.21.0

